### ISSUE TITLE
--holdout_after in single pass misses by 1 & breaks --progress

### ISSUE THREAD

**Issue by arielf** [VowpalWabbit/vowpal_wabbit#1608]
Trying to reproduce the issue reported by ayushpandey on the mailing list.

Noticed two possibly related issues as described in the title.
I'm guessing `--holdout_after` is supposed to imply `--holdout_period 1` unless otherwise specified.

Use a source tree data-set (`test/train-sets/0002.dat`) so anyone can reproduce the below.

Notice how the progress output stops the moment holdout kicks in (they should be independent).  Also notice the side-effect of partially fixing the problem (in 1st pass) when switching from one pass to 2 pass. Lastly notice the off-by-one example in starting holdout.

```
#
# Show that train-set has 1000 lines
#
$ wc -l test/train-sets/0002.dat
1000 test/train-sets/0002.dat

#
# Try to hold-out after 500 examples: printing progress every 100 lines
# Result: progress stops when hold-out is supposed to take effect
# Also: it misses by 1: stops learning after 499 rather than after 500
#
$ vw -k -P 100 --holdout_after 500 test/train-sets/0002.dat
...
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
0.020368 0.020368          100          100.0   0.5266   0.5147       15
0.012680 0.004993          200          200.0   0.4325   0.4097       15
0.009549 0.003287          300          300.0   0.5299   0.5367       15
0.007654 0.001969          400          400.0   0.4884   0.4973       15

finished run
number of examples = 499
weighted example sum = 499.000000
weighted label sum = 261.098202
average loss = 0.003018 h
best constant = 0.523243
total feature number = 7482
```

Adding `-c --passes 2` makes it print the progress between 500 & 1000 (1st pass), but breaks later when 2nd pass starts (no more progress printing):

```
$ vw -k -P 100 --holdout_after 500 test/train-sets/0002.dat -c --passes 2
...
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
0.020368 0.020368          100          100.0   0.5266   0.5147       15
0.012680 0.004993          200          200.0   0.4325   0.4097       15
0.009549 0.003287          300          300.0   0.5299   0.5367       15
0.007654 0.001969          400          400.0   0.4884   0.4973       15
0.003018 0.003018          500          500.0   0.5211   0.5286       15 h
0.003018  unknown          600          600.0   0.5634   0.5731       15 h
0.003018  unknown          700          700.0   0.5203   0.5037       15 h
0.003018  unknown          800          800.0   0.4048   0.4362       15 h
0.003018  unknown          900          900.0   0.5339   0.4960       15 h

finished run
number of examples per pass = 499
passes used = 2
weighted example sum = 998.000000
weighted label sum = 522.196404
average loss = 0.001531 h
best constant = 0.523243
total feature number = 14964
```

**Comment by Ayush-iitkgp** [comment]
Thank you @arielf, @JohnLangford you might be interested :)

**Comment by JohnLangford** [comment]
Interesting.  Thanks @arielf for the careful investigation.

**Comment by yannstad** [comment]
I fixed the counting issue (see PR #1844), that explains the off-by-one example issue.
The other bug I see is the "unknown" value in the "since last" average loss field.

@arielf you wrote that "Adding -c --passes 2 makes it print the progress between 500 & 1000 (1st pass), but breaks later when 2nd pass starts (no more progress printing)"
Actually what is printed here is the progress between 0 and 500 (2nd pass), not the progress between 500 & 1000 (1st pass). So it does not look like a bug.

On the "todo" side, the documentation about `--holdout_*` options, including details on the console output and the meaning of `h`, can clearly be improved.

**Comment by yannstad** [comment]
Thank you very much @arielf for the reproduction details. I have added these 2 scenarios in RunTests.

**Comment by arielf** [comment]
@yannstad thanks for the off-by-one fix.

Just tried to verify latest code. Things look not yet fully fixed. It may be a separate/additional bug than the one you fixed.

The data-set has 1000 examples. We do 2 passes + print a progress line every 100 examples, so I expect progress output keep going (every 100 examples) till we reach 2000 like it does in other cases when holdout isn't involved.

I still see some unexpected things in latest (2019-04-13) head.

Here's how to reproduce (and ideally add tests for these cases)

## Compare:

### Single pass.  Everything works as expected: progress gets printed every 100 lines till 1000.
```
$ build/vowpalwabbit/vw -k -P 100 test/train-sets/0002.dat       
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
using no cache
Reading datafile = test/train-sets/0002.dat
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
0.020368 0.020368          100          100.0   0.5266   0.5147       15
0.012680 0.004993          200          200.0   0.4325   0.4097       15
0.009549 0.003287          300          300.0   0.5299   0.5367       15
0.007654 0.001969          400          400.0   0.4884   0.4973       15
0.006393 0.001347          500          500.0   0.5113   0.5266       15
0.005476 0.000891          600          600.0   0.6330   0.6054       15
0.004771 0.000541          700          700.0   0.3748   0.3338       15
0.004254 0.000634          800          800.0   0.5115   0.4930       15
0.003851 0.000627          900          900.0   0.5768   0.5553       15
0.003499 0.000336         1000         1000.0   0.5588   0.5587       15

finished run
number of examples = 1000
weighted example sum = 1000.000000
weighted label sum = 526.517586
average loss = 0.003499
best constant = 0.526518
total feature number = 14996
```

### Now switch to 2 passes (no `--holdout _after`). Progress goes till 1800 (why not 2000?):
```
$ build/vowpalwabbit/vw -k -P 100  test/train-sets/0002.dat -c --passes 2 
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
decay_learning_rate = 1
creating cache_file = test/train-sets/0002.dat.cache
Reading datafile = test/train-sets/0002.dat
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
0.021428 0.021428          100          100.0   0.5642   0.5903       15
0.012679 0.003931          200          200.0   0.5867   0.5728       15
0.009705 0.003756          300          300.0   0.5895   0.3837       15
0.007773 0.001975          400          400.0   0.5274   0.5257       15
0.006367 0.000743          500          500.0   0.5693   0.5509       15
0.005457 0.000909          600          600.0   0.5792   0.5621       15
0.004805 0.000892          700          700.0   0.5690   0.5493       15
0.004289 0.000675          800          800.0   0.5332   0.5325       15
0.003855 0.000389          900          900.0   0.5558   0.5328       15
0.002078 0.002078         1000         1000.0   0.5642   0.5573       15 h
0.001905 0.000159         1100         1100.0   0.5867   0.5817       15 h
0.001782 0.000426         1200         1200.0   0.5895   0.5726       15 h
0.001708 0.000804         1300         1300.0   0.5274   0.5138       15 h
0.001609 0.000312         1400         1400.0   0.5693   0.5939       15 h
0.001520 0.000265         1500         1500.0   0.5792   0.5455       15 h
0.001442 0.000276         1600         1600.0   0.5690   0.5537       15 h
0.001378 0.000342         1700         1700.0   0.5332   0.5317       15 h
0.001310 0.000140         1800         1800.0   0.5558   0.5410       15 h
<<<<< 1st bug: should also print 1900, and 2000 but they are missing
finished run
number of examples per pass = 900     <<< second bug: this should be 1000
passes used = 2
weighted example sum = 1800.000000   <<< this should be 2000
weighted label sum = 946.682633
average loss = 0.000336 h
best constant = 0.525935
total feature number = 26992
```

### And when we add `--holdout_after 500` progress stops after 1000, which is a bit worse:
```
build/vowpalwabbit/vw -k -P 100 --holdout_after 500 test/train-sets/0002.dat -c --passes 2 
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
decay_learning_rate = 1
creating cache_file = test/train-sets/0002.dat.cache
Reading datafile = test/train-sets/0002.dat
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
0.020368 0.020368          100          100.0   0.5266   0.5147       15
0.012680 0.004993          200          200.0   0.4325   0.4097       15
0.009549 0.003287          300          300.0   0.5299   0.5367       15
0.007654 0.001969          400          400.0   0.4884   0.4973       15
0.006393 0.001347          500          500.0   0.5113   0.5266       15
0.003059 0.003059          600          600.0   0.5266   0.5545       15 h
0.003059  unknown          700          700.0   0.4325   0.4185       15 h
0.003059  unknown          800          800.0   0.5299   0.5362       15 h
0.003059  unknown          900          900.0   0.4884   0.4934       15 h
0.003059  unknown         1000         1000.0   0.5113   0.5074       15 h

finished run
number of examples per pass = 500
passes used = 2
weighted example sum = 1000.000000
weighted label sum = 523.218958
average loss = 0.001528 h
best constant = 0.523219
total feature number = 14994
```

Bottom line is that we still have a bug in `--progress` when more than 1 pass is involved.
And we seem to have an additional independent bug when `--holdout_after` is added.

Thanks again for looking into this!

**Comment by yannstad** [comment]
This behavior is expected, even though it may be surprising or counter intuitive.

**2nd scenario with 2 passes and no `--holdout _after`**
When you run several passes, vw automatically switches to 'over-fit avoidance' mode by holding-out 10% of the examples.
This is documented in the [tutorial](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Tutorial).
That is why only 1800 examples out of 2000 are used for training.
This also explains why the letter `h` (like holdout) is added after line 1000 and below.

**3rd scenario with 2 passes and `--holdout_after 500`**
The 3rd column is the total number of training examples seen so far.

During the 1st pass, the console output is:
```
0.020368 0.020368          100          100.0   0.5266   0.5147       15
0.012680 0.004993          200          200.0   0.4325   0.4097       15
0.009549 0.003287          300          300.0   0.5299   0.5367       15
0.007654 0.001969          400          400.0   0.4884   0.4973       15
0.006393 0.001347          500          500.0   0.5113   0.5266       15
```
The first 500 examples are used for training and produce an output line.
The next 500 examples are kept for testing (holdout) and does not produce any output.

During the 2nd pass, the console output is:
```
0.003059 0.003059          600          600.0   0.5266   0.5545       15 h
0.003059  unknown          700          700.0   0.4325   0.4185       15 h
0.003059  unknown          800          800.0   0.5299   0.5362       15 h
0.003059  unknown          900          900.0   0.4884   0.4934       15 h
0.003059  unknown         1000         1000.0   0.5113   0.5074       15 h
```
The first 500 examples are used for training and produce an output line.
The next 500 examples are kept for testing (holdout) and does not produce any output.

**Comment by marco-rossi29** [comment]
I believe that @yannstad is correct. I think @arielf was under the wrong assumption
`I'm guessing --holdout_after is supposed to imply --holdout_period 1 unless otherwise specified.`

In VW, you either do holdout_after or holdout_period. The two cannot be mixed.

In conclusion, I think VW is behaving as expected but the expected behavior is not clear from the documentation. I feel we should improve the documentation to address this, but no code change is necessary.

**Comment by yannstad** [comment]
I agree.
Also, another thing that should be improved in the documentation is how the average loss is computed for lines without an `h` and lines with an `h`.

**Comment by arielf** [comment]
Thanks @yannstad & @marco-rossi29 for the explanation.
I still don't feel we are doing the best we can in the present code.

There seems to be an arbitrary (and not sensible) link between `--holdout...` and `--progress`.
 
If I ask for 2-passes on a 1000 example data-set, I expect to have overall 2000 examples processed (regardless of holdout or not).

Independently, if I ask for a progress line every 100 examples, I expect to see a progress line printed every 100 examples (holdout or not), even if the example is held-out, and not learned from.  This is independent of whether the example is printed with the `h` suffix, or even if it has an unknown label which we sometimes print as 'unknown'.

The suppression of the progress line in an unexpected way when holdout is in effect (multi-pass or not) seems just wrong.

IOW: shouldn't  `--progress/-P` be independent of any other option?
